### PR TITLE
Remove `StringOffsetTrait` and `BinaryOffsetTrait`

### LINKED ISSUE #1644: Remove `StringOffsetTrait` and `BinaryOffsetTrait`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
`StringOffsetTrait` and `BinaryOffsetTrait` are redundant. We should delete them.
See the discussion in #1638.

### PR DESCRIPTION
**Pull request by HaoYang670**
# Which issue does this PR close?

Closes #1644.

# Rationale for this change

# What changes are included in this PR?

1. Remove `StringOffsetTrait` and `BinaryOffsetTrait`
2. Add `get_data_type` function for `StringArray` and `BinaryArray` (unfortunately, this function cannot be `const` because `OffsetSizeTrait::is_large` is not `const`)

# Are there any user-facing changes?
Yes. Delete 2 public traits.

### CODE REVIEW

**Review by viirya — COMMENTED** [review]
Hmm, why don't we just make a `const` `is_large` in `OffsetSizeTrait` simply?

```rust
pub trait OffsetSizeTrait {
  const IS_LARGE: bool;
}
impl OffsetSizeTrait for i32 {
    const IS_LARGE: bool = false;
}

impl OffsetSizeTrait for i64 {
    const IS_LARGE: bool = true
}
```

**Review by viirya — COMMENTED** [review]
Looks good to me. Just one question above.

**Comment by HaoYang670** [comment]
> Hmm, why don't we just make a `const` `is_large` in `OffsetSizeTrait` simply?
> 
> ```rust
> pub trait OffsetSizeTrait {
>   const IS_LARGE: bool;
> }
> impl OffsetSizeTrait for i32 {
>     const IS_LARGE: bool = false;
> }
> 
> impl OffsetSizeTrait for i64 {
>     const IS_LARGE: bool = true
> }
> ```
Strongly agree with you! 
We could create a follow-up issue to track this.

File: arrow/src/array/array_binary.rs [hunk 1]
```diff
@@ -35,29 +35,23 @@ use crate::error::{ArrowError, Result};
...
-pub struct GenericBinaryArray<OffsetSize: BinaryOffsetSizeTrait> {
+pub struct GenericBinaryArray<OffsetSize: OffsetSizeTrait> {
     data: ArrayData,
     value_offsets: RawPtrBox<OffsetSize>,
     value_data: RawPtrBox<u8>,
 }
 
-impl<OffsetSize: BinaryOffsetSizeTrait> GenericBinaryArray<OffsetSize> {
+impl<OffsetSize: OffsetSizeTrait> GenericBinaryArray<OffsetSize> {
+    pub fn get_data_type() -> DataType {
```
Comment (tustvold) [inline arrow/src/array/array_binary.rs:47]:
```suggestion
    pub const fn get_data_type() -> DataType {
```

File: arrow/src/array/array_string.rs [hunk 2]
```diff
@@ -27,31 +27,25 @@ use crate::buffer::Buffer;
...
-pub struct GenericStringArray<OffsetSize: StringOffsetSizeTrait> {
+pub struct GenericStringArray<OffsetSize: OffsetSizeTrait> {
     data: ArrayData,
     value_offsets: RawPtrBox<OffsetSize>,
     value_data: RawPtrBox<u8>,
 }
 
-impl<OffsetSize: StringOffsetSizeTrait> GenericStringArray<OffsetSize> {
+impl<OffsetSize: OffsetSizeTrait> GenericStringArray<OffsetSize> {
+    pub fn get_data_type() -> DataType {
```
Comment (tustvold) [inline arrow/src/array/array_string.rs:41]:
```suggestion
    pub const fn get_data_type() -> DataType {
```

**Review by tustvold — APPROVED** [review]
Minor nit you can take or leave, will merge later today.

Thanks :+1:

File: arrow/src/array/array_binary.rs (hunk 1 above)
Comment (tustvold) [reply arrow/src/array/array_binary.rs:47]: Edit: just saw your comment

**Comment by tustvold** [comment]
Going to get this in as the merge conflict potential is high

File: arrow/src/array/array_binary.rs (hunk 1 above)
Comment (HaoYang670) [reply arrow/src/array/array_binary.rs:47]: Will do this in #1658. Thank you for your review @viirya @tustvold !!

**Comment by alamb** [comment]
arrow 14 is shaping up for another great release 🎉

**Pull request merged**